### PR TITLE
C# package & distribtest fixes

### PR DESCRIPTION
**Pull request by jtattermusch**
The 1.6.0 package works fine, but:
- fixing a few broken distribtests 
- fixing System.Threading.ThreadPool warning
- bump protobuf dependency to 3.4.0 to stay in sync with submodule

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @jskeet

**Comment by jtattermusch** [comment]
C# artifact build here: https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/762/

**Comment by jskeet** [comment]
You're a star, thanks :)

File: tools/dockerfile/distribtest/csharp_ubuntu1604_x64/Dockerfile [hunk 1]
```diff
@@ -17,11 +17,16 @@ FROM ubuntu:16.04
...
     ca-certificates-mono \
-    nuget
+    nuget \
+    && apt-get clean
 
 # make sure we have nuget 2.12+ (in case there's an older cached docker image)
 RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
 RUN echo "deb http://download.mono-project.com/repo/debian wheezy main" | tee /etc/apt/sources.list.d/mono-xamarin.list
-RUN apt-get update && apt-get install -y nuget
+RUN apt-get update && apt-get install -y nuget && apt-get clean
```
Comment (apolcyn) [inline tools/dockerfile/distribtest/csharp_ubuntu1604_x64/Dockerfile]: nit: Can we add the mono-xamarin apt source as the first thing in this docker file, and avoid installing nuget for a second time here?

Comment (jtattermusch) [reply tools/dockerfile/distribtest/csharp_ubuntu1604_x64/Dockerfile]: I wanted to install  the mono version that comes natively with ubuntu16.04.
The nuget upgrade is only a fix applied on top of the vanilla mono installation (and I'm hoping that one day we will be able to get rid of it).

**Comment by apolcyn** [comment]
is the c# package build failure on linux new? https://grpc-testing.appspot.com/job/gRPC_build_packages/737/platform=linux/testReport/junit/(root)/tasks/build_package_csharp_package_dotnetcli_linux/

**Comment by jtattermusch** [comment]
I'll will investigate the `Health.cs(44,31): error CS0012: The type 'IEquatable<>' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [/var/local/git/grpc/src/csharp/Grpc.HealthCheck/Grpc.HealthCheck.csproj]`  error.

**Comment by jtattermusch** [comment]
@jskeet  solving the problem with `Health.cs(44,31): error CS0012: The type 'IEquatable<>' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [/var/local/git/grpc/src/csharp/Grpc.HealthCheck/Grpc.HealthCheck.csproj]` actually turns out to be quite tricky.
Google.Protobuf is targetting net451 and so seems the generated code (after bumping protobuf from v3.3.0 to v3.4.0). That in turn makes it difficult to compile some of my .csproj because they contain generated protobuf code (some are there just for testing - like Grpc.Examples.*,  but there are other packages that actually get shipped as nuget packages Grpc.HealthCheck  and Grpc.Reflection - and they are currently targeting net45, same as all other gRPC projects).

**Comment by jskeet** [comment]
Eek - that's odd. Google.Protobuf really shouldn't target net451.

Will file a bug, and let's see if we can get a 3.4.1 out targeting net45.

**Comment by jskeet** [comment]
Filed https://github.com/google/protobuf/issues/3595

Won't be able to try it today, but will hopefully be able to do so tomorrow. Apologies for this (it looks like it was my fault in Google.Protobuf).

**Comment by jtattermusch** [comment]
@jskeet Thanks for filing the issue. To move forward with this PR, I removed the version bump from v3.3.0 to v3.4.0 (v3.3.0 seems to work well as it targets net45). Once we have protobuf v3.4.1, I'll bump the version separately.

**Comment by jtattermusch** [comment]
Started artifact build here: https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/803/

**Comment by jtattermusch** [comment]
@apolcyn  PTAL.

**Comment by jskeet** [comment]
Yup, completely agree with using 3.3.0 as a workaround. I don't think any of the new features in 3.4.0 are critical here.

**Comment by jtattermusch** [comment]
Known issues:
https://github.com/grpc/grpc/issues/11109
https://github.com/grpc/grpc/issues/12407
https://github.com/grpc/grpc/issues/12106
https://github.com/grpc/grpc/issues/12208
https://github.com/grpc/grpc/issues/12408

**Pull request merged**